<!DOCTYPE html>
<html>
<head>
  <title>Annotations, external data | Ideogram</title>
  <style>body {font: 12px Arial; line-height: 19.6px;}</style>
  <script type="text/javascript" src="../dist/js/ideogram.min.js"></script>
</head>
<body>
  <h1>Annotations, external data | Ideogram</h1>
  <a href=".">Back to overview</a> |
  <a href="annotations_tracks.html">Previous</a> |
  <a href="annotations_file_url.html">Next</a>
  <div id="container"></div>
  <script type="text/javascript">

  /**
  * This function gets genomic coordinate data for ACMG genes from NCBI EUtils.
  * EUtils docs: https://www.ncbi.nlm.nih.gov/books/NBK25500/
  * Ideogram.js then draws those genes as annotations on chromosomes.
  */
  function getAndDrawAcmgGenes() {

    var acmgGenes, gene, i, annots, annot,
        geneClause, geneID, geneIDs, topWeight,
        eutils, esearch, esummary, url, defaultParams;

    geneIDs = [];
    annots = [];

    acmgGenes = [
      "APC", "MYH11", "ACTA2", "MYLK", "TMEM43", "DSP", "PKP2", "DSG2", "DSC2",
      "BRCA1", "BRCA2", "SCN5A", "RYR2", "LMNA", "MYBPC3", "COL3A1", "GLA",
      "APOB", "LDLR", "MYH7", "TPM1", "MYBPC3", "PRKAG2", "TNNI3", "MYL3",
      "MYL2", "ACTC1", "RET", "PCSK9", "TNNT2", "TP53", "TGFBR1", "TGFBR2",
      "TGFBR1", "TGFBR2", "SMAD3", "KCNQ1", "KCNH2", "SCN5A", "MLH1", "MSH2",
      "MSH6", "PMS2", "RYR1", "CACNA1S", "FBN1", "TGFBR1", "MEN1", "RET",
      "MUTYH", "NF2", "SDHD", "SDHAF2", "SDHC", "SDHB", "STK11", "MUTYH",
      "PTEN", "RB1", "TSC1", "TSC2", "VHL"
    ];

    acGenes = {"keys":["chr","start","length","color"], "annots":[
    {"chr":"1","annots":[["virtual1",30000000,6000000,"rgba(255, 0, 0, 0.65)"], ["virtual2",40000000,5000000,"rgba(0, 255, 0, 0.65)"]]},
    {"chr":"2","annots":[["nsv531656",40738282,17125539,"rgba(255, 0, 0, 0.65)"], ["2",30000000,9000000,"rgba(255, 0, 0, 0.65)"]]},
    {"chr":"20","annots":[["observation 1",30954171,5816,"rgba(255, 0, 0, 0.65)"]]},
    {"chr":"3","annots":[[" ",30000000,9000000,"rgba(0, 255, 0, 0.65)"], ["virtual3",33000000,3000000,"rgba(255, 0, 0, 0.65)"]]},
    {"chr":"9","annots":[["a duplicate name",120000000,9000000,"rgba(0, 255, 0, 0.65)"], ["a duplicate name",12000000,12000000,"rgba(255, 0, 0, 0.65)"]]},
    {"chr":"foo", "annots": []},
    {"chr":"X","annots":[["observation 2",47422351,6720,"rgba(0, 0, 255, 0.65)"]]}]}



    // Batch request all ACMG genes
    //geneClause = "(" + acmgGenes.join("[symbol] OR ") + "[symbol])";

    topWeight = 0;
    weights = [];

    //url = esearch + "&term=Homo-sapiens[Organism] AND " + geneClause;
    //d3.json(url, function(error, data) {
      //geneIDs = data.esearchresult.idlist;
      //console.log(geneIDs)

      // Batch request genomic coordinates for all ACMG genes
      //url = esummary + "&id=" + geneIDs.join(",");
      d3.tsv('for_visuals.csv', function(error, data) {
        console.log(data[0].start);
        for (var i = 0; i < data.length; i++) { //data.length
          var gene, chr, start, stop, pheno;
          geneInfo = data[i];

          gene = geneInfo.gene;
          chr = geneInfo.chromosome;
          start = Number(geneInfo.start);
          //console.log(start)
          stop = Number(geneInfo.end);
          pheno = Number(geneInfo.phenotype);

          //if (result.weight > topWeight) {
          //  topWeight = result;
          //}

          ////
          // This is the annotations API
          ////
          annots.push({
            "name": gene, // required, but unused
            "chr": chr, // required
            "start": start, // required
            "stop": stop, // required
            "pheno": pheno // required
          });

        }

        // NCBI assigns each gene a weight based on how well it is characterized.
        // http://www.ncbi.nlm.nih.gov/books/NBK3841/#EntrezGene.Sort_by
        // Gene weight is a rough proxy for general biomedical relevance.
        // Higher weight, more important.
        //annots.sort(function(a, b) {
        //  return b.weight - a.weight;
        //});

        // This block highlights the 10 most important ACMG genes.
        for (i = 0; i < annots.length; i++) {
          //color = (i < 10) ? "#F00" : "#FAA";
          //color =  "#FAA"
          //annots[i].shape = "triangle";
          //annots[i].color = color;

          // BRCA1 is the most widely-recognized gene name.  Make it stand out.
          if (annots[i].pheno === 0) {
            annots[i].shape = "triangle";
            //console.log("hurray")
            annots[i].color = "#6BAED6"; // BREAST
          }
          if (annots[i].pheno === 1) {
            annots[i].shape = "triangle";
            annots[i].color = "#74c476"; // IDC
          }
          if (annots[i].pheno === 2) {
            annots[i].shape = "triangle";
            annots[i].color = "#fd8d3c"; //orange, MDLC
          }
          if (annots[i].pheno === 3) {
            annots[i].shape = "triangle";
            annots[i].color = "#9e9ac8"; //orange, IDC
          }
          if (annots[i].pheno === -1) {
            annots[i].shape = "triangle";
            annots[i].color = "#ff5252"; // Multiple Membership
          }

        }

        ideogram.drawAnnots(annots);

      });

  }

  var d3 = Ideogram.d3;

  var config = {
    organism: "human",
    orientation: "vertical",
    resolution: 550,
    chrHeight: 300,
    annotationHeight: 4,
    onLoad: getAndDrawAcmgGenes
  };

  var ideogram = new Ideogram(config);

  </script>
</body>
</html>
